### PR TITLE
[SYCL] Re-allow auto& parameter in range parallel_for

### PR DESCRIPTION
**Pull request by jzc**
Before #10864, our header implementation allowed kernels with an auto& parameter, e.g. `q.parallel_for(1, [=](auto &id) {})` , but strictly speaking, this would not follow the criteria in [4.9.4.2.2. parallel_for invoke](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_parallel_for_invoke). After that PR, this behavior unintentionally became disallowed. This PR restores the old behavior while disallowing it under `-DSYCL2020_CONFORMANT_APIS`.

### CODE REVIEW

**Comment by aelovikov-intel** [comment]
Is the focus on `auto` or `&` here?

**Review by cperkinsintel — APPROVED** [review]

**Comment by jzc** [comment]
> Is the focus on `auto` or `&` here?

`&`, kernels with just `auto` should still work as expected

File: sycl/test/basic_tests/handler/parallel_for_arg_restrictions.cpp [hunk 1]
```diff
@@ -71,6 +71,18 @@ int main() {
...
+    CGH.parallel_for(sycl::range{1}, [=](auto &) {});
+  });
+  Q.submit([&](sycl::handler &CGH) {
+    // expected-error@sycl/handler.hpp:* {{Kernel should be invocable with a sycl::item}}
+    CGH.parallel_for(sycl::range{1, 1}, [=](auto &) {});
+  });
+  Q.submit([&](sycl::handler &CGH) {
+    // expected-error@sycl/handler.hpp:* {{Kernel should be invocable with a sycl::item}}
+    CGH.parallel_for(sycl::range{1, 1, 1}, [=](auto &) {});
+  });
```
Comment (aelovikov-intel) [inline sycl/test/basic_tests/handler/parallel_for_arg_restrictions.cpp:85]: Please move `auto &` tests next to `auto`. If those don't exist, then we need to add them here.

**Comment by againull** [comment]
@jzc  Could you please address comment from @aelovikov-intel so that we can merge this.

File: sycl/test/basic_tests/handler/parallel_for_arg_restrictions.cpp (hunk 1 above)
Comment (jzc) [reply sycl/test/basic_tests/handler/parallel_for_arg_restrictions.cpp:85]: Testing `parallel_for` with an auto parameter is already tested by `handler_generic_lambda_interface.cpp` https://github.com/intel/llvm/blob/be8e031cd078ea0298a773aa5fc743c18e50fe97/sycl/test/basic_tests/handler/handler_generic_lambda_interface.cpp#L16-L25

**Comment by hvdijk** [comment]
Is it intentional that `parallel_for` now requires being invokable with both `item<Dims>` and `item<Dims, false>`? This prevents lambdas from taking `id<Dims>` since that is only convertible to `item<Dims>`, not from `item<Dims, false>`, but the part of the SYCL specification that you link to says that the intent is to allow taking `id<Dims>`. This causes a large number of errors when attempting to build the SYCL CTS.

**Comment by jzc** [comment]
@hvdijk That was a mistake, the `&&` should have been an `||`. Should be fixed by https://github.com/intel/llvm/pull/11625.

**Pull request merged**